pkg/cloud/cloud.go

```diff
@@ -576,15 +590,22 @@ func (c *cloud) DetachDisk(ctx context.Context, volumeID, nodeID string) error {
 		return err
 	}
 
-	// TODO: check if attached
+	disk, err := c.GetDiskByID(ctx, volumeID)
+	if err != nil {
+		if errors.Is(err, ErrNotFound) {
+			return status.Errorf(codes.NotFound, "volume %q not found", volumeID)
+		}
+		return fmt.Errorf("could not get volume with ID %q: %w", volumeID, err)
+	}
+
 	device, err := c.dm.GetDevice(instance, volumeID)
 	if err != nil {
 		return err
 	}
 	defer device.Release(true)
 
-	if !device.IsAlreadyAssigned {
-		klog.InfoS("DetachDisk: called on non-attached volume", "volumeID", volumeID)
+	if !isDiskAttached(disk, nodeID) && !device.IsAlreadyAssigned {
+		return nil
```

Comment:
I am VERY worried that this is fundamentally unsafe, especially in a scenarios involving parallelism or multiple instances of the controller (which is, by my understanding, *legal* according to the CSI spec)

Can we please add a justification explaining why this is safe as a comment in the code, both to
1. Prove to us, on this PR, that this change is actually safe, and
2. To convey the reasoning for that to future maintainers

Reply:
Added a comment. 

This early exit condition is safe because if the disk is not attached to the node and the device is not assigned there is nothing more to do in `DetachDisk`. This operation is idempotent, which means it can be repeated without changing the final outcome, thus even in a scenario with parallelism or multiple controller instances this function is safe due to the idempotency.

There is one potential race condition: its possible that between the time we check if the disk is attached and when `DetachVolumeWithContext` is called another process has already detached the volume. In the worst case this would simply result in an extra unnecessary API call. This scenario is handled gracefully by the error checks following the `DetachVolumeWithContext` call.

> multiple instances of the controller (which is, by my understanding, legal according to the CSI spec)

By my understanding more than one controller running concurrently is not supported, see https://github.com/kubernetes-sigs/gcp-compute-persistent-disk-csi-driver/issues/516#issuecomment-635453521 has there been a change or update to the CSI spec that allows for this?

Reply:
> This early exit condition is safe because if the disk is not attached to the node and the device is not assigned there is nothing more to do in DetachDisk

My point is that I don't think the code as written actually verifies that "the disk is not attached to the node"

I'm not talking about k8s, I'm talking about CSI

Reply:
Let's take the extra API call and attempt to detach regardless of the disk's perceived attachment status to address eventual consistency.